packages/truffle-compile/parser.js

```diff
@@ -50,19 +47,6 @@ module.exports = {
       ({ message }) => !message.includes(preReleaseCompilerWarning)
     );
 
-    // If the import error key is not found, we must not have an import error.
-    // This means we have a *different* parsing error which we should show to the user.
-    // Note: solc can return multiple parsing errors at once.
-    const nonImportErrors = errors.filter(
-      ({ formattedMessage }) => !formattedMessage.includes(importErrorKey)
-    );
-
-    // Should we try to throw more than one? (aside; we didn't before)
-    if (nonImportErrors.length > 0) {
-      throw new CompileError(nonImportErrors[0].formattedMessage);
-    }
-
-    // Now, all errors must be import errors.
```

Comment:
@CruzMolina Sadly I *think* this might not work as expected. It looks like if a file has both an import error and a syntax error, solc exits with the syntax error and doesn't report the invalid import. It's like it won't consider the imports until other basic stuff is taken care of. To see an example of this try compiling
```solidity
pragma solidity 1.2.3;

import "I/am/not/a/file";

contract Problems {
  constructor() public {
  }

  just some words;

}
```
and logging the solc output at this location in the code. Then comment out `just some words` and try again. 

If the syntax error doesn't get thrown here it means that files with invalid imports would cycle through the profiler, which could be bad. Conversely, if you throw here the compilers really have to match.

Reply:
Ah, so it looks like `solc-js` will throw/return the first SyntaxError it encounters.

Thanks for pointing this out! I _think_ I have a solution 🤞 .

Reply:
@cgewecke I pushed some failsafe code to handle "fail even faster than bad imports" scenarios. After testing various scenarios locally things _appear_ fine, but I probably should add some Solidity Import scenario tests.

Reply:
One scenario I'm worried about is compiler mismatching. For example suppose the user has specified an `0.4.24` compiler in their config and the built-in/parsing compiler is `0.5.0`. The users code will have statements which are legal in `0.4.24` but errors in `0.5.0`. If truffle throws for these statements while imports parsing, the ability to choose any compiler is broken. 


> Ah, so it looks like solc-js will throw/return the first SyntaxError it encounters.

I'm confused by how solc behaves fwiw. To me it seems like:
+  it throws certain classes of order in sequence   
+ solc expects the user to resolve these different error classes *in a series of compilation passes*. 
+ import paths errors are not necessarily caught on the first pass. I *think* the code sample above is an example of that.

Does fail safe filtering address that if the file has both a (perceived) syntax error *and* a bad import path?

As an aside - it looks like frangio is fixing a problem with bonkers async whilsts (mea culpa) in #2009 that touches the same files as this PR - it might make sense to merge that to develop if it looks ok and rebase here since it goes nicely with the profiler simplification you're doing.

Reply:
Apologies, I've edited the above quite a bit - I find this error injection parsing strategy difficult to reason about.

Reply:
I concur with merging #2009 first

Reply:
@cgewecke I definitely hear your concern, can you provide another example? It appears that imaginary statement _is_ a valid ParserError in 0.4.24 (and appears to be a ParserError at least as far back as 0.4.0).

I've been trying to reproduce any kind of bug or red flag with the current solution and haven't come up with anything yet (throwing in "uncompilable" examples from Solidity's 0.5.0 Breaking Changes {https://solidity.readthedocs.io/en/v0.5.0/050-breaking-changes.html?highlight=breaking#interoperability-with-older-contracts} into the parser while configuring truffle to use older versions of solc doesn't appear to break anything).

FWIW, solc classifies the bad file imports truffle throws in _and_ any source code not conforming to Solidity language rules as a ParserError. It appears that actual wacky code takes precedence over bad file imports. For our purposes, I think all we care about for parser compilation is that the source code is linguistically legal before passing it on to the actual supplied compiler. We let that compiler handle all the SyntaxError, DeclarationError, TypeError, and Warning error types.

Reply:
@CruzMolina Ok I could be misunderstanding something. I'd like to see if I can create failing test case for this later today. Thanks for investigating.

Reply:
Np! Thanks for helping reason about this. In case I wasn't being clear, the parser injection hack _forces_ solc to fail fast, before even getting to version-specific errors. Bad file imports appear to be lower in priority compared to actual grammatical issues. Would love to know if this doesn't actually hold true! 🤞

Reply:
@CruzMolina I think #2011 is an example. Maybe.

Also had a thought today that if mixing compilers *is* a box of worms, one *could* use solidity-parser-antlr as the default but enable slow 'expedient hack' import parsing with the user's compiler via a flag or something. For the small % of cases where the JS parser lags the solc parser.  Probably have to explain this option pretty clearly to the user via the cli when they get syntax errors though.

Reply:
Thanks for the PR @cgewecke!

I _think_ the solution here is to only match failing imports and then filter out any non-import matches (instead of throwing to avoid any unintended compiler mixing effects).

To contextualize the broader situation a bit, we've been actively interested in removing `solc-js` from the bundle (Windows compatibility issues). Until we're able to do so, your issue makes a strong case for using the bundled `solc-js`. I think our preference at this point is to not introduce another dependency for parsing (would be nice if solc provided a parser for us).